main.py

```diff
@@ -127,10 +145,47 @@ def update_db(
     conn.commit()
     conn.close()
 
+# Calculate total cost and tokens for detailed report
+def calculate_cost_and_tokens(result):
+
+    total_embedding_cost = 0.0
+    total_embedding_tokens = 0
+    total_llm_cost = 0.0
+    total_llm_tokens = 0
+        
+    # Extract 'extraction_result' from the result
+    extraction_result = result.get("extraction_result", [])
+        
+    if extraction_result:
+        extraction_data = extraction_result[0].get("result", "")
+     
+        # If extraction_data is a string, attempt to parse it as JSON
+        if isinstance(extraction_data, str):
+            try:
+                extraction_data = json.loads(extraction_data) if extraction_data else {}
+            except json.JSONDecodeError:
+                logger.warning("Failed to decode JSON for extraction data; defaulting to empty dictionary.")
+                extraction_data = {}
+
+       
+        metadata = extraction_data.get("metadata", {})
+        embedding_llm = metadata.get("embedding", [])
+        extraction_llm = metadata.get("extraction_llm", [])
+
+        # Calculate total cost
+        total_embedding_cost += sum(float(item.get("cost_in_dollars", "0")) for item in embedding_llm)
+        total_llm_cost += sum(float(item.get("cost_in_dollars", "0")) for item in extraction_llm)
+
+        # Calculate total tokens
+        total_embedding_tokens += sum(item.get("embedding_tokens", 0) for item in embedding_llm)
+        total_llm_tokens += sum(item.get("total_tokens", 0) for item in extraction_llm)
```

Comment:
Hi I think we should consider writing it as normal for loop. Be cause we are unnecessarily looping through same variables twice here. For small data that might be okay. But with larger data we might see performance impact. Better approach would be to have 2 for loops. One for embedding and one for llm. Then get token and cost together